core/src/dsnp/graph_page.rs

```diff
@@ -1,44 +1,158 @@
 #![allow(dead_code)] // todo: remove
 use crate::{
 	dsnp::{api_types::*, dsnp_types::*},
+	graph::updates::{UpdateEvent, UpdateTracker},
 	util::time::time_in_ksecs,
 };
-use std::collections::HashMap;
+use anyhow::{Error, Result};
+use std::collections::{HashMap, HashSet};
+
+use super::{
+	compression::{CompressionBehavior, DeflateCompression},
+	encryption::EncryptionBehavior,
+	schema::SchemaHandler,
+};
 
 /// Graph page structure
 #[derive(Debug, Clone, PartialEq)]
 pub struct GraphPage {
+	/// Page ID
+	page_id: PageId,
+	/// Privacy type of owning graph
+	privacy_type: PrivacyType,
+	/// Current content hash of page as retrieved from chain
+	content_hash: u32,
 	/// List of PRIds
 	prids: Vec<DsnpPrid>,
 	/// List of connections
 	connections: Vec<DsnpGraphEdge>,
 }
 
+/// Conversion for Public Graph
+impl TryFrom<&PageBlob> for GraphPage {
+	type Error = Error;
+	fn try_from(PageBlob { content_hash, payload, page_id, .. }: &PageBlob) -> Result<Self> {
+		let DsnpUserPublicGraphChunk { compressed_public_graph } =
+			DsnpUserPublicGraphChunk::try_from(payload.as_slice())?;
+		let uncompressed_chunk = DeflateCompression::decompress(&compressed_public_graph)?;
+		Ok(Self {
+			page_id: *page_id,
+			privacy_type: PrivacyType::Public,
+			content_hash: *content_hash,
+			prids: Vec::new(),
+			connections: SchemaHandler::read_inner_graph(&uncompressed_chunk)?,
+		})
+	}
+}
+
+/// Conversion for Private Graph
+impl<E: EncryptionBehavior> TryFrom<(&PageBlob, &Vec<KeyPair<E>>)> for GraphPage {
+	type Error = Error;
+	fn try_from(
+		(PageBlob { content_hash, payload, page_id, .. }, keys): (&PageBlob, &Vec<KeyPair<E>>),
+	) -> Result<Self> {
+		let DsnpUserPrivateGraphChunk { key_id, encrypted_compressed_private_graph, prids } =
+			DsnpUserPrivateGraphChunk::try_from(payload.as_slice())?;
+		let mut decrypted_chunk: Option<Vec<u8>> = None;
+
+		// First try the key that was indicated in the page
+		if let Some(indicated_key) = keys.iter().find(|k| k.key_id == key_id) {
```

Comment:
I think if we can not decrypt using the key mentioned we should fail the process? Right? Decryption might work with a wrong key but the decrypted data would be completely wrong and cause further problem down the line.

Reply:
The only historical checks should be done for PRI calculations